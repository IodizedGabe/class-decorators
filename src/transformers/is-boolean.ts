import { composePropertyDecorators, IsBooleanOptions } from '@hippo-oss/dto-decorators';
import { Transform, TransformFnParams, Type } from 'class-transformer';

import { createBasePropertyDecorators } from './base';

export function IsBoolean(options: IsBooleanOptions = {}): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // convert string to boolean
        Type(() => Boolean),

        // convert 'false' and '0' to false
        Transform(({ obj, key, value }: TransformFnParams): unknown => {
            /* NB: by the time this function is called, the @Type() decorator will have already converted
             * the orginal value to boolean, so we must recover the original value; in additional, decorator
             * order has no impact on this behavior because class-transformer saves decorator logic as
             * reflect-metadata and evaluates it using its own ordering logic.
             */
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            const originalValue = obj[key] as unknown;

            if (typeof originalValue === 'string') {
                if (originalValue === '0') {
                    return false;
                }
                if ('false'.localeCompare(originalValue, undefined, { sensitivity: 'base' }) === 0) {
                    return false;
                }
            }

            return value as unknown;
        }),
    ]);
}
