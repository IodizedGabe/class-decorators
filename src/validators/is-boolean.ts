import { composePropertyDecorators, IsBooleanOptions } from '@hippo-oss/dto-decorators';
import { IsBoolean as IsBooleanDecorator } from 'class-validator';

import { createBasePropertyDecorators } from './base';

export function IsBoolean(options: IsBooleanOptions = {}): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate data as a boolean
        IsBooleanDecorator({ each: !!options.isArray }),
    ]);
}
