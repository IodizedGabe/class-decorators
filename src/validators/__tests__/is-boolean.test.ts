import { validateSync } from 'class-validator';

import { IsBoolean } from '../is-boolean';

describe('IsBoolean', () => {
    it('allows optional boolean to be omitted', () => {
        class Fixture {
            @IsBoolean({
                optional: true,
            })
            property!: boolean;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable boolean to be omitted', () => {
        class Fixture {
            @IsBoolean({
                nullable: true,
            })
            property!: boolean;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required boolean to be omitted', () => {
        class Fixture {
            @IsBoolean({})
            property!: boolean;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non-boolean', () => {
        class Fixture {
            @IsBoolean({})
            property!: boolean;
        }

        const obj = new Fixture();
        obj.property = 'string' as unknown as boolean;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows boolean', () => {
        class Fixture {
            @IsBoolean({})
            property!: boolean;
        }

        const obj = new Fixture();
        obj.property = true;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
});
