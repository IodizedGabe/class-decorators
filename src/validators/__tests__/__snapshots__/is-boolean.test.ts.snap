// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsBoolean does not allow non-boolean 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "property must be a boolean value",
    },
    "property": "property",
    "target": Fixture {
      "property": "string",
    },
    "value": "string",
  },
]
`;

exports[`IsBoolean does not allow required boolean to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "property must be a boolean value",
      "isDefined": "property should not be null or undefined",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
